libexec/rtld-elf/aarch64/reloc.c

```diff
@@ -588,18 +603,16 @@ reloc_non_plt(Obj_Entry *obj, Obj_Entry *obj_rtld, int flags,
 		 */
 		case R_MORELLO_CAPINIT:
 		case R_MORELLO_GLOB_DAT:
-			if (process_r_cheri_capability(obj, ELF_R_SYM(rela->r_info), lockstate,
-			    flags, where, rela->r_addend) != 0)
+			if (process_r_cheri_capability(obj,
+			    ELF_R_SYM(rela->r_info), lockstate, flags,
+			    where, rela->r_addend) != 0)
 				return (-1);
 			break;
 		case R_MORELLO_RELATIVE:
-#ifdef __CHERI_PURE_CAPABILITY__
-			init_cap_from_fragment(obj->relocbase, obj->text_rodata_cap,
-			    rela->r_offset, rela->r_addend);
-#else
-			init_cap_from_fragment(obj->relocbase, obj->relocbase,
-			    rela->r_offset, rela->r_addend);
-#endif
+			init_cap_from_fragment(where, data_cap,
+			    text_rodata_cap,
+			    (Elf_Addr)(uintptr_t)obj->relocbase,
```

Comment:
Since we only support address interpretation nowadays, this could be `(Elf_Addr)obj->relocbase`?

Reply:
`-Wcapability-to-integer-cast` would give a warning still, though that's not enabled by `-Wall -Wextra -Wcheri`.

Reply:
(which is in `-Wcheri-pedantic`)

Reply:
Hmm I forgot about that warning. The intention was that casts to types tagged as address types (which Elf_Addr/ptraddr_t should be) don't trigger it. But I can't remember if I implemented that correctly.

Reply:
https://cheri-compiler-explorer.cl.cam.ac.uk/z/1fbrc5

Reply:
Uh, https://cheri-compiler-explorer.cl.cam.ac.uk/z/5doEo1, I meant to have MIPS on the other panel. The other reason I like this pattern is it works equally if you use `uintcap_t` and a capability in hybrid, both in that you get the right (non-CToPtr) behaviour and that you don't get the additional (non-pedantic) warnings that brings on top of the `-Wcapability-to-integer-cast` one.